### PR TITLE
fix: RuleTester JSDoc and deprecations

### PR DESCRIPTION
**Pull request by xkr47**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

Without this fix, I get a misguided warning in IntelliJ IDEA that the "new-style" rule defintition on line 14 from [this example in the documentation](https://eslint.org/docs/latest/extend/custom-rule-tutorial#step-6-write-the-test) does not match the type `Function`  — it should accept both the legacy `Function` and new style, object-based `Rule` definitions.

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Tell us about your environment (`npx eslint --env-info`):**
* Node version: v18.13.0
* npm version: v8.19.3
* Local ESLint version: v8.47.0 (Currently used)
* Global ESLint version: Not found
* Operating System: linux 6.4.11-100.fc37.x86_64

**What parser are you using (place an "X" next to just one item)?**

[X] `Default (Espree)`
[ ] `@typescript-eslint/parser`
[ ] `@babel/eslint-parser`
[ ] `vue-eslint-parser`
[ ] `@angular-eslint/template-parser`
[ ] `Other`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>
See https://eslint.org/docs/latest/extend/custom-rule-tutorial
</details>

**What did you do? Please include the actual source code causing the issue.**

See https://eslint.org/docs/latest/extend/custom-rule-tutorial

**What did you expect to happen?**

The example code should not generate this kind of warning in IntelliJ IDEA.

**What actually happened? Please include the actual, raw output from ESLint.**

IntelliJ IDEA (and other JSDoc-supporting source code editors) warns about an incorrect function argument when the JSDoc definition actually only allows the deprecated old-style rule definition.

#### What changes did you make? (Give an overview)

Updated the JSDoc to support the new-style rule definition as well. Also updated a neighbouring function the same way and copy-pasted the check for legacy rule definitions from the run() function source code.

#### Is there anything you'd like reviewers to focus on?

I have not checked any other files than the one I modified for similar issues.

### CODE REVIEW

**Review by mdjermanovic — COMMENTED** [review]
Changes LGTM. Can you also fix JSDoc for `FlatRuleTester#run`?

https://github.com/eslint/eslint/blob/1fbb3b0b477c814c0d179564fe495f4c50a451e9/lib/rule-tester/flat-rule-tester.js#L449

**Comment by xkr47** [comment]
Could you verify that the change I did for defineRule() (the second commit in this PR) is correct?

**Comment by xkr47** [comment]
I am asking because of how run() and defineRule() are defined in the typescript typings for the same: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/utils/src/ts-eslint/RuleTester.ts#L152

**Comment by xkr47** [comment]
> Changes LGTM. Can you also fix JSDoc for `FlatRuleTester#run`?
> 
> https://github.com/eslint/eslint/blob/1fbb3b0b477c814c0d179564fe495f4c50a451e9/lib/rule-tester/flat-rule-tester.js#L449

I have added the JSDoc fix. However, since:
1. the `emitLegacyRuleAPIWarning()` helper is only available in rule-tester.js, and
2. flat-rule-tester.js seems to be available only through https://github.com/eslint/eslint/blob/main/lib/unsupported-api.js

.. so I was thinking maybe it is not necessary to announce interface deprecations in unsupported apis. The flat-rule-tester.js DOES have tests though, so your call.

File: lib/rule-tester/flat-rule-tester.js
Comment (mdjermanovic) [inline lib/rule-tester/flat-rule-tester.js:35]:
```suggestion
/** @typedef {import("../shared/types").Rule} Rule */
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by mdjermanovic** [comment]
> Could you verify that the change I did for defineRule() (the second commit in this PR) is correct?

Looks good to me. typescript-eslint types seem specific for their typescript-eslint plugin rules.

> I have added the JSDoc fix. However, since:
> 
> 1. the `emitLegacyRuleAPIWarning()` helper is only available in rule-tester.js, and
> 2. flat-rule-tester.js seems to be available only through https://github.com/eslint/eslint/blob/main/lib/unsupported-api.js
> 
> .. so I was thinking maybe it is not necessary to announce interface deprecations in unsupported apis.

It isn't necessary to emit deprecation warnings in FlatRuleTester (https://github.com/eslint/eslint/pull/16063#issuecomment-1193192107).

**Comment by xkr47** [comment]
Well spotted, apologizes for that. Fixed.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! Leaving open in case someone else wants to review it before merging.

**Review by snitin315 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
